README.md
PyAjam
======

[AJAM](http://www.voip-info.org/wiki/view/Aynchronous+Javascript+Asterisk+Manager+%28AJAM%29) (Asynchronous Javascript Asterisk Manager) is the new interface build upon HTTP to interact with Asterisk. 
AJAM was Introduced in asterisk 1.4. Returned datas are mostly formatted in XML.

[PyAjam](http://devedge.bour.cc/wiki/PyAjam) is a python library allowing
programs to interact with an asterisk server using the AJAM interface, in
a pythonic way.


###Requirements:
* None 

###Compatility:
*	python 2.[5-7]

###Installation
	easy_install PyAjam

or

	wget http://devedge.bour.cc/resources/pyajam/src/pyajam.latest.tar.gz
	tar xvf pyajam.latest.tar.gz
	cd reblok-* && ./setup.py install

Features
--------

* retrieve basic informations (sip/iax2 peers, sip peer details, sip registry)
* catch asterisk events
* information presented in a pythonic way (dict)
* automatically reconnect to the asterisk server if connection lost
* compatible with asterisk 1.4 and 1.6

Documentation
-------------

You can found PyAjam documentation at [http://devedge.bour.cc/resources/pyajam/doc/index.html](http://devedge.bour.cc/resources/pyajam/doc/index.html)

Example
-------

	>>> from pyajam import Pyajam

	>>> ajam = Pyajam(server='192.168.0.10', username='mspencer', password='*rocks!')
	>>> if not ajam.login():
	>>>		print "Invalid login"
	>>>		sys.exit(1)

	>>> # display list of peers (SIP and IAX2)
	>>> print ajam.peers())
	>>> # display peer 101 attributes
	>>> peer  = ajam.sippeer('101'))

	>>> # screenprint events
	>>> def ajam_event_listener(data):
	>>>		print data

	>>> ajam.waitevent(async=False, callback=ajam_event_listener)

About
-----

*PyAjam* is licensed under GNU GPL v3.
It is developped by Guillaume Bour <guillaume@bour.cc>


CHANGES

0.1 - initial release

0.1.1 - docs
	add README.md, MANIFEST.in, AUTHORS, CHANGES and COPYING 

0.2 - maintenance (lstep)
    . replace use of 4suite-xml by xml.etree
    . add unittests
    . add callbacks for lost/recovered connections notification



tests/test_ajam.py
from nose import SkipTest
from nose.tools import assert_equal

from pyajam import Pyajam

class TestPyajam:
    def setUp(self):
      pass

    def tearDown(self):
      pass

    def test_ajam_unify_xml(self):
        input = '''<ajax-response>
                  <response type='object' id='unknown'><generic response='Success' message='Waiting for Event...' /></response>
                  <response type='object' id='unknown'><generic event='PeerStatus' privilege='system,all' peer='SIP/601' peerstatus='Registered' /></response>
                  <response type='object' id='unknown'><generic event='WaitEventComplete' /></response>
                </ajax-response> '''
        expected = [{u'peer': u'SIP/601', u'peerstatus': u'Registered', u'event': u'PeerStatus', u'privilege': u'system,all'}, {u'event': u'WaitEventComplete'}]

        ajam = Pyajam()
        
        #raise SkipTest
        assert_equal(expected, ajam._unify_xml(input))


tests/test_decoding.py
# -*- coding: utf8 -*-
# vim:syntax=python:sw=4:ts=4:expandtab
"""
    pyajam, Python Asterisk AJAM binding
    Copyright (C) 2009-2012  Guillaume Bour <guillaume@bour.cc>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import unittest
import os
import os.path

from pyajam import Pyajam


def fake_query(_path, _action):
    """We fake the code querying Asterisk AJAM interface

        _qq() read the file traces/${ersion}/${action}
        this file must contains what AJAM would return (including status code and HTTP headers)
    """
    def _qq(mode, action, args={}):        
        headers = dict()
        content = None

        if action == 'command':
            action = args['command']

        with open(os.path.join(_path, action)) as f:
            # first line is HTTP 
            f.readline()

            while True:
                l = f.readline()
                if l == '\r\n':
                    break

                (key, value) = l[:-1].split(':', 1)
                if key.lower() == 'set-cookie':
                    key = key.lower()
                headers[key] = value.strip()

            content = f.read()
        return (headers, content)
    return _qq

class TestDecoding(unittest.TestCase):
    def __init__(self, name, version=None):
        super(TestDecoding, self).__init__(name)
        self.version = version
        self.path = os.path.join(os.getcwd(), "traces", version)

    def setUp(self):
        self.ajam = Pyajam()
        self.ajam._query = fake_query(self.path, self._testMethodName.split('_')[1])

    def tearDown(self):
        self.ajam = None

    def test_login(self):
        self.assertTrue(self.ajam.login())
        self.assertEqual(self.ajam.version(), self.version)
        self.assertEqual(self.ajam._sessionid_, '3ac1e27b')

    def test_sippeers(self):
        peers = self.ajam.sippeers()
        self.assertEqual(len(peers), 2)
        self.assertTrue(peers[0].get('objectname', 'none') in ['101','102'])

    def test_iaxpeers(self):
        peers = self.ajam.iaxpeers()
        self.assertEqual(len(peers), 1)
        self.assertEqual(peers[0].get('objectname','none'), 'demo')

    def test_peers(self):
        peers = self.ajam.peers()
        self.assertEqual(len(peers), 3)
        self.assertEqual(len([x for x in peers if x['channeltype'] == 'SIP']), 2)
        self.assertEqual(len([x for x in peers if x['channeltype'] == 'IAX2']), 1)

    def test_sipregistry(self):
        regs = self.ajam.sipregistry()
        self.assertEqual(len(regs), 1)
        self.assertEqual(regs[0]['username'], 'myISP')

    def test_peer(self):
        peer = self.ajam.sippeer('101')
        self.assertTrue(isinstance(peer, dict))
        self.assertEqual(peer.get('name',None), '101')
        self.assertEqual(peer.get('status',None), 'Unmonitored')


if __name__ == '__main__':
    testnames =  unittest.TestLoader().getTestCaseNames(TestDecoding)

    for vers in ['1.4']:
        print "** with asterisk 1.4 **"

        suite = unittest.TestSuite()
        for name in testnames:
            suite.addTest(TestDecoding(name, vers))

        unittest.TextTestRunner(verbosity=2).run(suite)



pyajam/pyajam.py
# -*- coding: utf8 -*-
#
# vim:syntax=python:sw=4:ts=4:expandtab
"""
    pyajam, ...
    Copyright (C) 2009-2012  Guillaume Bour <guillaume@bour.cc>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
__author__  = "Guillaume Bour <guillaume@bour.cc>"
__status__  = "developpement"
__version__ = "0.2"
__date__    = "2011/09/08"

import sys, urllib2, re, logging, time
# Only on python 2.5 & superior
from xml.etree import ElementTree

# only on python 2.5 & superior
if not hasattr(urllib2, 'quote'):
  import urllib
  urllib2.quote = urllib.quote


class Pyajam:

  def __init__(self, server='localhost', port=8088, path='/asterisk', 
               username='admin', password='admin', autoconnect=True, 
               asraw=False):
    """Pyajam initialisation.

     Parameters:
      * **server**: adress or IP of the asterisk server to connect to (default= *localhost*),
      * **port**: ajam port (as defined in `asterisk/http.conf <http://www.voip-info.org/wiki/index.php?page=Asterisk+config+http.conf>`_),
      * **path**: url path to send ajam queries,
      * **username**: ajam login username (as defined in `asterisk/manager.conf <http://www.voip-info.org/tiki-index.php?page=Asterisk%20config%20manager.conf>`_),
      * **password**: ajam login password,
      * **autoconnect**: Pyajam automatically (re)connect to AJAM server,
      * **raw**: return values as raw (e.g not pythonified)
    """
    self.url          = "http://%s:%d%s" % (server, port, path)
    self.username     = username
    self.password     = password
    self.autoconnect  = autoconnect
    self.unify        = not asraw

    self.connexion_status = 'DISCONNECTED' # 'DISCONNECTED', 'CONNECTED','LOST_CONNEXION'
    self.errback      = None
    self.recoverback  = None

    self._sessionid_  = ''
    self._version_    = None
    self._waitevt__run = False

  def _query(self, mode, action, args={}):
    """Query Asterisk AJAM service.

    """
    qargs = '&'.join(map(lambda x: "%s=%s" % (x, urllib2.quote(args[x])), args))
    if len(qargs) > 0:
      qargs = '&' + qargs

    logging.debug("GET %s/%s?action=%s%s" % (self.url, mode, action, qargs))
    req = urllib2.Request("%s/%s?action=%s%s" % (self.url, mode, action, qargs))
    req.add_header('Cookie', "mansession_id=\"%s\"" % self._sessionid_)
    try:
      f = urllib2.urlopen(req)    
      data = f.read()
      logging.debug(data)
    except Exception, e:
      logging.error(e)
      if self.connexion_status == 'CONNECTED' and self.errback:
        # We WERE connected, but not anymore, so if there is a errback, run it
        self.errback()
      self.connexion_status = 'LOST_CONNEXION'
      return (None, None)

    if self.autoconnect and \
       ('Authentication Required' in data or 'Permission denied' in data):
      logging.debug('trying to reconnect...')
      if not self.login():
        if self.connexion_status == 'CONNECTED' and self.errback:
          # We WERE connected, but not anymore, so if there is a errback, run it
          self.errback()
        self.connexion_status = 'LOST_CONNEXION'
        return (None, None)

      req.add_header('Cookie', "mansession_id=\"%s\"" % self._sessionid_)
      try:
        f = urllib2.urlopen(req)
        data = f.read()
      except Exception, e:
        if self.connexion_status == 'CONNECTED' and self.errback:
          # We WERE connected, but not anymore, so if there is a errback, run it
          self.errback()
        self.connexion_status = 'LOST_CONNEXION'
        return (None, None)
    

    if self.connexion_status == 'LOST_CONNEXION' and self.recoverback:
        self.connexion_status = 'CONNECTED'
        # We have a recover, say it !
        self.recoverback()

    self.connexion_status = 'CONNECTED'
    return (f.info(), data)

  def _unify_xml(self, raw, normalizer=None, attribute='event'):
    """Convert Asterisk ajam XML message to python dictionary.
    """
    doc = ElementTree.fromstring(raw)
    datas = []

    for elem in doc.iter(tag='generic'):
      if attribute in elem.keys():
        # We have the good shape :-)
        attrs = {}
        for key in elem.keys():
          attrs[key] = elem.get(key)
        if normalizer:
          attrs = normalizer(attrs)
        if attrs:
          datas.append(attrs)

    return datas

  def _unify_raw(self, raw, regex, normalizer, datas=None):
    """Convert Asterisk raw message (old manager API) to python dictionary.
    """
    if regex is None:
      return raw

    if type(regex) == str:
      regex = re.compile(regex, re.M)

    if datas is None:
        datas=[]

    for m in regex.finditer(raw):
      obj = m.groupdict()

      if normalizer:
        obj = normalizer(obj)

      if obj:
        if isinstance(datas, list):
          datas.append(obj)
        elif isinstance(datas, dict):
          datas.update(obj)

    return datas

  def login(self):
    """Login to Asterisk AJAM service.
     Use the username & password provided at class instantiation.

     Returns **True** if login success, else **False**
    """
    (info, data) = self._query('manager', 'login', 
      {'username': self.username, 'secret': self.password})
    if not info or not "Authentication accepted" in data:
      self._sessionid_ = ''
      return False

    # setting session
    #print f.info()
    self._sessionid_ = info['set-cookie'].split(';')[0].split('=')[1][1:-1]

    # extraction asterisk version
    if not info.has_key('Server') or \
    not info['Server'].startswith('Asterisk/'):
      logging.error("login:: not logging to an asterisk server")
      return False

    self._version_ = info['Server'].split('/')[1][:3]
#   print 'version=', self._version_
    if self._version_ not in ['1.4', '1.6']:
      logging.error("login:: Unmanaged %s asterisk version" % self._version_)
      return 'False'
    
    return True

  def version(self):
    """Return Asterisk version of server you are connected to.
     Values are **None** (not connected), **1.4** or **1.6**.
    """
    return self._version_

  def sippeers(self):
    """Query SIP peers definition.
     Return list of peers (each peer is represented as a dictionary).

     Useful items are:
       * **objectname**, the peer name
       * **status**, with *OK* value if peer is connected
       * **ipaddress**, device IP

     New fields in asterisk 1.6:
       * **textsupport**

     ::

     # sippeers() output sample
     >>> import pprint
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> pprint.pprint(ajam.sippeers())
     [ { u'acl'           : u'no',
         u'channeltype'   : u'SIP',
         u'chanobjecttype': u'peer',
         u'dynamic'       : u'yes',
         u'event'         : u'PeerEntry',
         u'ipaddress'     : u'-none-',
         u'ipport'        : u'0',
         u'natsupport'    : u'no',
         u'objectname'    : u'102',
         u'realtimedevice': u'no',
         u'status'        : u'UNKNOWN',
         u'videosupport'  : u'no',
         u'textsupport'   : u'no'},
       { u'acl'           : u'no',
         u'channeltype'   : u'SIP',
         u'chanobjecttype': u'peer',
         u'dynamic'       : u'yes',
         u'event'         : u'PeerEntry',
         u'ipaddress'     : u'127.0.0.1',
         u'ipport'        : u'5061',
         u'latency'       : u'1',
         u'natsupport'    : u'no',
         u'objectname'    : u'101',
         u'realtimedevice': u'no',
         u'status'        : u'OK',
         u'videosupport'  : u'no',
         u'textsupport'   : u'no'}]
    """
    (info, data) = self._query('mxml', 'sippeers')
    if not info:
      return False

    if self.unify:
      def _normalize(row):
        if row['event'] != 'PeerEntry':
          return None

        # when OK, status = 'OK (13 ms)' per example (with latency in parenthesis)
        if row['status'][0:2] == 'OK':
          row[u'latency'] = row['status'][4:-4]
          row[u'status']  = u'OK'

        return row

      data = self._unify_xml(data, _normalize)
      #data = self._unify_xml(data, lambda r: r['event'] == 'PeerEntry' and r or None)
    return data

  def iaxpeers(self):
    """Query IAX peers definition.
     Return list of peers (each peer is represented as a dictionary).

     Useful items are:
       * **objectname**, the peer name
       * **status**, with *OK* value if peer is connected
       * **ipaddress**, device IP

     New fields in asterisk 1.6:
       * **chanobjecttype**
       * **encryption**
       * **event**
       * **trunk**

     Notes:
       * **username** field is truncated if longer than 10 characters

     ::

     # iaxpeers() output sample
     >>> import pprint
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> pprint.pprint(ajam.IAXpeers())
     [ { u'channeltype'   : u'IAX2',
         u'dynamic'       : u'no',
         u'ipaddress'     : u'216.207.245.47',
         u'ipport'        : u'4569',
         u'objectname'    : u'demo',
         u'username'      : u'asterisk',
         u'status'        : u'Unmonitored',
         u'chanobjecttype': u'peer',
         u'encryption'    : u'no',
         u'event'         : u'PeerEntry',
         u'trunk'         : u'no'}]


    """
    mode = 'rawman'
    if self._version_ == '1.6':
      mode = 'mxml'
    (info, data) = self._query(mode, 'iaxpeers')

    if not info:
      return False

    if self.unify:
      def _normalize(row):
        row[u'channeltype'] = u'IAX2'
    
        if row['dynamic'] == 'D':
          row['dynamic'] = u'yes'
        else:
          row['dynamic'] = u'no'
    
        # when OK, status = 'OK (13 ms)' per example
        if row['status'][0:2] == 'OK':
          row[u'latency'] = row['status'][4:-4]
          row[u'status']  = 'OK'

        name = row['objectname'].split('/')
        if len(name) > 1:
          row[u'objectname']  = name[0]
          row[u'username']    = name[1]
        else:
          row[u'username']    = u''

        return row

    if self._version_ == '1.6':
      data = self._unify_xml(data, _normalize)
    else:
      data = self._unify_raw(data, 
        '^(?P<objectname>[^\s]+)\s+(?P<ipaddress>[0-9.]*)\s+\\((?P<dynamic>S|D)\\)\s+([0-9.]+)\s+(?P<ipport>\d+)\s+(?P<status>.*?)\s+$',
        _normalize
      )
    return data

  def peers(self):
    """Query both IAX and SIP peers.
     Return list of peers (each peer is represented as a dictionary).
     Use **channeltype** to distinguish between *sip* and *iax2* devices.
     ::

     # peers() output sample
     >>> import pprint
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> pprint.pprint(ajam.peers())
     [ { u'channeltype'   : u'IAX2',
         u'chanobjecttype': u'peer',
         u'dynamic'       : u'no',
         u'encryption'    : u'no',
         u'event'         : u'PeerEntry',
         u'ipaddress'     : u'-none-',
         u'ipport'        : u'0',
         u'objectname'    : u'iax1',
         u'status'        : u'Unmonitored',
         u'trunk'         : u'no',
         u'username'      : u''},
       { u'acl'           : u'no',
         u'channeltype'   : u'SIP',
         u'chanobjecttype': u'peer',
         u'dynamic'       : u'yes',
         u'event'         : u'PeerEntry',
         u'ipaddress'     : u'127.0.0.1',
         u'ipport'        : u'5061',
         u'latency'       : u'1',
         u'natsupport'    : u'no',
         u'objectname'    : u'101',
         u'realtimedevice': u'no',
         u'status'        : u'OK',
         u'textsupport'   : u'no',
         u'videosupport'  : u'no'}]
    """
    peers = self.sippeers()
    peers.extend(self.iaxpeers())

    return(peers)

  def sipregistry(self):
    """Query SIP devices registration status
     (mainly used for trunks).

     New fields in asterisk 1.6:
       * **event**
       * **registrationtime**

     ::

     # sipregistry() output sample
     >>> import pprint
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> pprint.pprint(ajam.sipregistry())
     [ { u'event': u'RegistryEntry',
         u'host'            : u'mysipprovider.com',
         u'port'            : u'5060',
         u'refresh'         : u'120',
         u'registrationtime': u'0',
         u'state'           : u'Unregistered',
         u'username'        : u'1234'}]
    """
    if self._version_ == '1.6':
      def _normalize(row):
        if row['event'] != 'RegistryEntry':
          return None
        return row

      (info, data) = self._query('mxml', 'sipshowregistry')
      if not info:
        return False

      data = self._unify_xml(data, _normalize)
    else:
      data = self.command('sip show registry',
        '^(?P<host>[^\s]+):(?P<port>[0-9]+)\s+(?P<username>[^\s]+)\s+(?P<refresh>[0-9.]+)\s+(?P<state>[^\s]+).*$'
      )

    return data

#     raw command: sip show peer <peername>
#     Note: keys are lowered
  def sippeer(self, peername):
    """Query details about a SIP peer.
     Comparing to sippeers(), here you have got detailed informations about the peer.

     New fields in asterisk 1.6:
       * a lot of fields have been renamed or added between 1.4 and 1.6

     ::

     # sippeer() output sample
     >>> import pprint
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> pprint.pprint(ajam.sippeer('101'))
     { 'acl'          : 'No',
       'addr->ip'     : '127.0.0.1 Port 5061',
       'ama flags'    : 'Unknown',
       'auto-framing' : 'No',
       'call limit'   : '0',
       'callerid'     : '"" <>',
       'callgroup'    : '',
       'callingpres'  : 'Presentation Allowed, Not Screened',
       'canreinvite'  : 'Yes',
       'codec order'  : '(none)',
       'codecs'       : '0x8000e (gsm|ulaw|alaw|h263)',
       'context'      : 'default',
       'def. username': '101',
       'defaddr->ip'  : '0.0.0.0 Port 5060',
       'dtmfmode'     : 'rfc2833',
       'dynamic'      : 'Yes',
       'expire'       : '1801',
       'insecure'     : 'no',
       'language'     : '',
       'lastmsg'      : '0',
       'lastmsgssent' : '32767/65535',
       'mailbox'      : '',
       'maxcallbr'    : '384 kbps',
       'md5secret'    : '<Not set>',
       'name'         : '101',
       'nat'          : 'RFC3581',
       'overlap dial' : 'No',
       'pickupgroup'  : '',
       'promiscredir' : 'No',
       'reg. contact' : 'sip:101@127.0.0.1:5061',
       'secret'       : '<Not set>',
       'send rpid'    : 'No',
       'sip options'  : '(none)',
       'status'       : 'OK (1 ms)',
       'subscr.cont.' : '<Not set>',
       'subscriptions': 'Yes',
       't38 pt udptl' : 'No',
       'tohost'       : '',
       'transfer mode': 'open',
       'trust rpid'   : 'No',
       'user=phone'   : 'No',
       'useragent'    : 'Twinkle/1.4.2',
       'video support': 'No',
       'vm extension' : 'asterisk'}

    """
    if self._version_ == '1.6':
      (info, data) = self._query('mxml', 'sipshowpeer', {'peer': peername})
      if not info:
        return False

      data = self._unify_xml(data, attribute='response')
    else:
      def _normalizer(row):
        key = row['key'].strip()
        if key in ['Response', 'Privilege']:
          return None

        if key == '* Name': 
          key = 'Name'

        row[key.lower()] = row['value'].strip()

        del row['key']
        del row['value']
        return row

      data = self.command('sip show peer ' + peername,
        '^(?P<key>.*?):(?P<value>.*)$',
        _normalizer, {}
      )
    return data
    

  def command(self, command, regex=None, normalizer=None, unifyin=None):
    """Execute an Asterisk command.
      command is always returned in raw mode from asterisk.
      *command* is the asterisk command as typed in asterisk console,
      *regex* is a python regular expression use to parse command returned value,
      *normalizer* is used to normalize returned results
      *unifyin* is either an empty list or dict, as you want to return a list or a dict

     ::

     # direct call command() output sample
     >>> import pprint, re
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> 
     >>> rx = re.compile('^(?P<key>[^\s]+)\s+(?P<value>.*?)\s+\s+.*$', re.M)
     >>> def module_show_normalize(row):
     ...   if row['key'] in ('--END', 'Response:', 'Privilege:', 'Module') or \\
     ...      row['value'] == 'modules':
     ...   return None
     ...
     ...   row[row['key']] = row['value']
     ...   del row['key']
     ...   del row['value']
     ...
     ...   return row
     ...
     >>> pprint.pprint(ajam.command('module show', rx, module_show_normalize))
     [{'res_features.so': 'Call Features Resource'},
      {'res_musiconhold.so': 'Music On Hold Resource'},
      {'pbx_config.so': 'Text Extension Configuration'},
      {'chan_sip.so': 'Session Initiation Protocol (SIP)'},
      {'app_ldap.so': 'LDAP directory lookup function for Aster 0'}]

    """
    (info, data) = self._query('rawman', 'command', {'command': command})
    if not info:
      return False

    if self.unify and regex:
      data = self._unify_raw(data, regex, normalizer, unifyin)
    return data

  def waitevent(self, async=False, callback=None, errback=None, recoverback=None):
    """Manage Asterisk events.
     
     Usage:
       * if *async* is **False** and *callback* is **None**, return the first eventlist received (*blocking*),
       * if *async* is **False** and *callback* is not **None**, execute the callback for 
         each received eventlist (*blocking*, e.g waitevent() never returns until calling *waitevent_stop()*),
       * if *async* is **True** (*callback* is required), execute the callback for each 
         eventlist (*non-blocking*, e.g waitevent() returns immediatly)

     **Callback function**:
        the callback function must take one argument, an eventlist (array of
        events, as dictionaries)

     ::

     # waitevent() output sample
     >>> import pprint, re
     >>> from pyajam import Pyajam
     >>> ajam = Pyajam(username='mspencer', password='*rocks!')
     >>> pprint ajam.waitevent(async=False)
     >>>
     >>> def ajam_event_listener(evtlist):
     >>>   pp.pprint(evtlist)
     >>> ajam.waitevent(async=False, callback=ajam_event_listener)
     [ { u'event': u'Reload',
         u'message': u'Reload Requested',
         u'privilege': u'system,all'},
        {u'event': u'WaitEventComplete'}]
    """
    if self._waitevt__run:
      raise Exception('Event manager is still running')

    if errback:
      self.errback = errback

    if recoverback:
      self.recoverback = recoverback

    if async:
      if not callback:
        raise ValueError('callback argument is required when async is true')

      from Queue import Queue
      import thread
      self.eventQ = Queue(0)
      thread.start_new_thread(self._waitevent, (callback,))
    else:
      return self._waitevent(callback)

  def _waitevent(self, callback):
    """Event async manager.
    """
    self._waitevt__run = True
    while self._waitevt__run:
      (info, data) = self._query('mxml', 'waitevent')
      if not info:
        time.sleep(1); continue

      if self.unify:
        data = self._unify_xml(data)
    
      if callback:
        try:
          callback(data)
        except Exception, e:
          print e
      else:
        self._waitevt__run = False
        return data     

  def waitevent_stop(self):
    """Stop event async manager.
    """
    self._waitevt__run = False


if __name__ == '__main__':
  logging.basicConfig(level=logging.ERROR)

  import pprint
  pp = pprint.PrettyPrinter(indent=2)

  ajam = Pyajam()
  if not ajam.login():
    print "Invalid login"
    sys.exit(1)

  print 'asterisk version=', ajam.version()
  # Get SIP peers list
  pp.pprint(ajam.sippeers())
  # Get IAX2 peers list
  pp.pprint(ajam.iaxpeers())
  # Get SIP+IAX2 peers list
  pprint.pprint(ajam.peers())
  # Get SIP registry
  pp.pprint(ajam.sipregistry())
  # Get SIP peer 101
  pp.pprint(ajam.sippeer('101'))

  # Get modules list
  rx = re.compile('^(?P<key>[^\s]+)\s+(?P<value>.*?)\s+\s+.*$', re.M)
  def module_show_normalize(row):
    if row['key'] in ('--END', 'Response:', 'Privilege:', 'Module') or row['value'] == 'modules':
      return None

    row[row['key']] = row['value']
    del row['key']
    del row['value']
  
    return row
  pprint.pprint(ajam.command('module show', rx, module_show_normalize))

  # Event handler
  def ajam_event_listener(data):
    print "event data >>>"
    pp.pprint(data)

  while True:
    pp.pprint(ajam.waitevent(False))
    time.sleep(1)

# ajam.waitevent(False, ajam_event_listener)

# ajam.waitevent(True, ajam_event_listener)
# while True:
#   time.sleep(1)
